general/mem_manager.hpp

```diff
@@ -26,28 +26,64 @@ namespace mfem
 /// Memory types supported by MFEM.
 enum class MemoryType
 {
-   HOST,      ///< Host memory; using new[] and delete[]
-   HOST_32,   ///< Host memory aligned at 32 bytes (not supported yet)
-   HOST_64,   ///< Host memory aligned at 64 bytes (not supported yet)
-   CUDA,      ///< cudaMalloc, cudaFree
-   CUDA_UVM   ///< cudaMallocManaged, cudaFree (not supported yet)
+   HOST,           ///< Host memory; using new[] and delete[]
+   HOST_32,        ///< Host memory; aligned at 32 bytes
+   HOST_64,        ///< Host memory; aligned at 64 bytes
+   HOST_DEBUG,     ///< Host memory; protected while in device mode
+   HOST_UMPIRE,    ///< Host memory; using Umpire
+   HOST_MANAGED,   ///< Host memory; using *MallocManaged, *Free
+   DEVICE,         ///< Device memory; using *Malloc, *Free
+   DEVICE_DEBUG,   ///< Device memory; protected while in host mode
+   DEVICE_UMPIRE,  ///< Device memory; using Umpire
+   DEVICE_MANAGED, ///< Device memory; using D2D copies only
+   SIZE            ///< Number of host and device memory types
 };
 
-/// Memory classes identify subsets of memory types.
-/** This type is used by kernels that can work with multiple MemoryType%s. For
-    example, kernels that can use CUDA or CUDA_UVM memory types should use
-    MemoryClass::CUDA for their inputs. */
+/// Size of the memory types, the host and device ones.
+constexpr int MemoryTypeSize = static_cast<int>(MemoryType::SIZE);
+constexpr int HostMemoryTypeSize = static_cast<int>(MemoryType::DEVICE);
+constexpr int DeviceMemoryTypeSize = MemoryTypeSize - HostMemoryTypeSize;
+
+/// Memory type names, used during Device:: configuration.
+extern const char *MemoryTypeName[MemoryTypeSize];
+
+/// Pre & post increment operators for the MemoryType.
+inline MemoryType& operator++(MemoryType &mt)
```

Comment:
Can you point me to where these increment and decrement operations are used?

Reply:
[Here](https://github.com/mfem/mfem/blob/bda839d3db666e1fa49c063618acdfe3873813f7/general/mem_manager.cpp#L595) and [here](https://github.com/mfem/mfem/blob/bda839d3db666e1fa49c063618acdfe3873813f7/tests/unit/general/test_mem.cpp#L25)

Reply:
Hmm, it seems a little redundant -- the loop variables can be simply `int` since the initialization and comparisons with the enum constants should work just fine (they are implicitly convertible to `int`). Casting `int` to `MemoryType` will have to be explicit but that seems simpler. Not a big issue but it could be a small simplification here.

Reply:
I had some warnings with `qtcreator`, but that can be simplified if wanted, sure.

Reply:
I'll leave this is up to you, @camierjs.

Reply:
`done`